astropy/table/tests/test_index.py

```diff
@@ -623,3 +623,12 @@ def test_nd_columun_as_index(masked):
         ValueError, match="Multi-dimensional column 'arr' cannot be used as an index."
     ):
         t.add_index("arr")
+
+
+def test_quantity_column_as_index():
+    # Regression test for https://github.com/astropy/astropy/issues/16036
+    t = QTable({"a": u.Quantity([0, 1], unit="m"), "b": [1, 2]})
```

Comment:
As we discussed on the issue, there are two different failure modes on `main` (for scalar/non-scalar `Quantity`), so I'd suggest testing both
```suggestion
@pytest.mark.parametrize("input_data", [[0], [0,1]])
def test_quantity_column_as_index(input_data):
    # Regression test for https://github.com/astropy/astropy/issues/16036
    t = QTable({"a": u.Quantity(input_data, unit="m"), "b": [1, 2]})
```

Reply:
(there's probably a more elegant way to parametrize it though)